src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/FixedSizeArrayBuilder.cs

```diff
@@ -0,0 +1,87 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Runtime.InteropServices;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+/// <summary>
+/// A bare-bones array builder, focused on the case of producing <see cref="ImmutableArray{T}"/>s where the final array
+/// size is known at construction time.  In the golden path, where all the expected items are added to the builder, and
+/// <see cref="MoveToImmutable"/> is called, this type is entirely garbage free.  In the non-golden path (usually
+/// encountered when a cancellation token interrupts getting the final array), this will leak the intermediary array
+/// created to store the results.
+/// </summary>
+/// <remarks>
+/// This type should only used when all of the following is true:
+/// <list type="number">
+/// <item>
+/// The number of elements is known up front, and is fixed.  In other words, it isn't just an initial-capacity, or a
+/// rough heuristic.  Rather it will always be the exact number of elements added.
+/// </item>
+/// <item>
+/// Exactly that number of elements is actually added prior to calling <see cref="MoveToImmutable"/>.  This means no
+/// patterns like "AddIfNotNull".
+/// </item>
+/// <item>
+/// The builder will be moved to an array (see <see cref="MoveToArray"/>) or <see cref="ImmutableArray{T}"/> (see <see
+/// cref="MoveToImmutable"/>).
+/// </item>
+/// </list>
+/// If any of the above are not true.  For example, the capacity is a rought hint, or the exact number of elements may
+/// not match the capacity specified, or if it's intended as a scratch buffer, and won't realize a final array, then
+/// <see cref="ArrayBuilder{T}.GetInstance(int, T)"/> should be used instead.
+/// </remarks>
+[NonCopyable]
+internal struct FixedSizeArrayBuilder<T>(int capacity)
+{
+    private T[] _values = new T[capacity];
+    private int _index;
+
+    public void Add(T value)
+        => _values[_index++] = value;
+
+    public void AddRange(ImmutableArray<T> values)
+    {
+        foreach (var v in values)
```

Comment:
>foreach (var v in values) [](http://example.com/codeflow?start=8&length=25)

Should we do any input validation before writing off the end of the array?

Reply:
sure.